From: Stained Glass OS <ke7oxh@gmail.com>
d3drm: mesh builder vertex and normal accessors, boxes and flags

IDirect3DRMMeshBuilder2 and 3 implement SetVertex/SetNormal (and the
array forms), GetVertex/GetNormal, Translate, GetBox, ReserveSpace,
GetGeometry for version 3, the perspective flag, the color source and the
texture topology.
---
--- a/dlls/d3drm/d3drm_private.h
+++ b/dlls/d3drm/d3drm_private.h
@@ -229,6 +229,9 @@ struct d3drm_mesh_builder
     struct mesh_material *materials;
     DWORD *material_indices;
     D3DRMRENDERQUALITY quality;
+    D3DRMCOLORSOURCE color_source;
+    BOOL perspective;
+    BOOL wrap_u, wrap_v;
 };
 
 struct mesh_group
--- a/dlls/d3drm/meshbuilder.c
+++ b/dlls/d3drm/meshbuilder.c
@@ -529,23 +529,56 @@ static HRESULT WINAPI d3drm_mesh_builder2_Scale(IDirect3DRMMeshBuilder2 *iface,
 static HRESULT WINAPI d3drm_mesh_builder2_Translate(IDirect3DRMMeshBuilder2 *iface,
         D3DVALUE tx, D3DVALUE ty, D3DVALUE tz)
 {
-    FIXME("iface %p, tx %.8e, ty %.8e, tz %.8e stub!\n", iface, tx, ty, tz);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder2(iface);
+    DWORD i;
 
-    return E_NOTIMPL;
+    TRACE("iface %p, tx %.8e, ty %.8e, tz %.8e.\n", iface, tx, ty, tz);
+
+    for (i = 0; i < mesh_builder->nb_vertices; ++i)
+    {
+#ifndef SG_MUTANT_DRMMB_NO_TRANSLATE
+        mesh_builder->vertices[i].x += tx;
+#endif
+        mesh_builder->vertices[i].y += ty;
+        mesh_builder->vertices[i].z += tz;
+    }
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder2_SetColorSource(IDirect3DRMMeshBuilder2 *iface, D3DRMCOLORSOURCE source)
 {
-    FIXME("iface %p, source %#x stub!\n", iface, source);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder2(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, source %#x.\n", iface, source);
+
+    if ((int)source != D3DRMCOLOR_FROMFACE && (int)source != D3DRMCOLOR_FROMVERTEX)
+        return D3DRMERR_BADVALUE;
+    mesh_builder->color_source = source;
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder2_GetBox(IDirect3DRMMeshBuilder2 *iface, D3DRMBOX *box)
 {
-    FIXME("iface %p, box %p stub!\n", iface, box);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder2(iface);
+    DWORD i;
 
-    return E_NOTIMPL;
+    TRACE("iface %p, box %p.\n", iface, box);
+
+    if (!box)
+        return D3DRMERR_BADVALUE;
+    memset(box, 0, sizeof(*box));
+    for (i = 0; i < mesh_builder->nb_vertices; ++i)
+    {
+        const D3DVECTOR *v = &mesh_builder->vertices[i];
+
+        if (!i || v->x < box->min.x) box->min.x = v->x;
+        if (!i || v->y < box->min.y) box->min.y = v->y;
+        if (!i || v->z < box->min.z) box->min.z = v->z;
+        if (!i || v->x > box->max.x) box->max.x = v->x;
+        if (!i || v->y > box->max.y) box->max.y = v->y;
+        if (!i || v->z > box->max.z) box->max.z = v->z;
+    }
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder2_GenerateNormals(IDirect3DRMMeshBuilder2 *iface)
@@ -557,9 +590,11 @@ static HRESULT WINAPI d3drm_mesh_builder2_GenerateNormals(IDirect3DRMMeshBuilder
 
 static D3DRMCOLORSOURCE WINAPI d3drm_mesh_builder2_GetColorSource(IDirect3DRMMeshBuilder2 *iface)
 {
-    FIXME("iface %p stub!\n", iface);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder2(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p.\n", iface);
+
+    return mesh_builder->color_source;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder2_AddMesh(IDirect3DRMMeshBuilder2 *iface, IDirect3DRMMesh *mesh)
@@ -604,10 +639,19 @@ static HRESULT WINAPI d3drm_mesh_builder2_AddFaces(IDirect3DRMMeshBuilder2 *ifac
 static HRESULT WINAPI d3drm_mesh_builder2_ReserveSpace(IDirect3DRMMeshBuilder2 *iface,
         DWORD vertex_count, DWORD normal_count, DWORD face_count)
 {
-    FIXME("iface %p, vertex_count %lu, normal_count %lu, face_count %lu stub!\n",
-            iface, vertex_count, normal_count, face_count);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder2(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, vertex_count %lu, normal_count %lu, face_count %lu.\n", iface, vertex_count, normal_count, face_count);
+
+    if (vertex_count > mesh_builder->vertices_size &&
+            !d3drm_array_reserve((void **)&mesh_builder->vertices, &mesh_builder->vertices_size, vertex_count,
+            sizeof(*mesh_builder->vertices)))
+        return E_OUTOFMEMORY;
+    if (normal_count > mesh_builder->normals_size &&
+            !d3drm_array_reserve((void **)&mesh_builder->normals, &mesh_builder->normals_size, normal_count,
+            sizeof(*mesh_builder->normals)))
+        return E_OUTOFMEMORY;
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder2_SetColorRGB(IDirect3DRMMeshBuilder2 *iface,
@@ -662,9 +706,13 @@ static HRESULT WINAPI d3drm_mesh_builder2_SetMaterial(IDirect3DRMMeshBuilder2 *i
 static HRESULT WINAPI d3drm_mesh_builder2_SetTextureTopology(IDirect3DRMMeshBuilder2 *iface,
         BOOL wrap_u, BOOL wrap_v)
 {
-    FIXME("iface %p, wrap_u %#x, wrap_v %#x stub!\n", iface, wrap_u, wrap_v);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder2(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, wrap_u %#x, wrap_v %#x.\n", iface, wrap_u, wrap_v);
+
+    mesh_builder->wrap_u = !!wrap_u;
+    mesh_builder->wrap_v = !!wrap_v;
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder2_SetQuality(IDirect3DRMMeshBuilder2 *iface,
@@ -678,25 +726,44 @@ static HRESULT WINAPI d3drm_mesh_builder2_SetQuality(IDirect3DRMMeshBuilder2 *if
 
 static HRESULT WINAPI d3drm_mesh_builder2_SetPerspective(IDirect3DRMMeshBuilder2 *iface, BOOL enable)
 {
-    FIXME("iface %p, enable %#x stub!\n", iface, enable);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder2(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, enable %#x.\n", iface, enable);
+
+    mesh_builder->perspective = !!enable;
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder2_SetVertex(IDirect3DRMMeshBuilder2 *iface,
         DWORD index, D3DVALUE x, D3DVALUE y, D3DVALUE z)
 {
-    FIXME("iface %p, index %lu, x %.8e, y %.8e, z %.8e stub!\n", iface, index, x, y, z);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder2(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, index %lu, x %.8e, y %.8e, z %.8e.\n", iface, index, x, y, z);
+
+#ifndef SG_MUTANT_DRMMB_NO_BOUNDS
+    if (index >= mesh_builder->nb_vertices)
+        return D3DRMERR_BADVALUE;
+#endif
+    mesh_builder->vertices[index].x = x;
+    mesh_builder->vertices[index].y = y;
+    mesh_builder->vertices[index].z = z;
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder2_SetNormal(IDirect3DRMMeshBuilder2 *iface,
         DWORD index, D3DVALUE x, D3DVALUE y, D3DVALUE z)
 {
-    FIXME("iface %p, index %lu, x %.8e, y %.8e, z %.8e stub!\n", iface, index, x, y, z);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder2(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, index %lu, x %.8e, y %.8e, z %.8e.\n", iface, index, x, y, z);
+
+    if (index >= mesh_builder->nb_normals)
+        return D3DRMERR_BADVALUE;
+    mesh_builder->normals[index].x = x;
+    mesh_builder->normals[index].y = y;
+    mesh_builder->normals[index].z = z;
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder2_SetTextureCoordinates(IDirect3DRMMeshBuilder2 *iface,
@@ -825,9 +892,11 @@ static D3DRMRENDERQUALITY WINAPI d3drm_mesh_builder2_GetQuality(IDirect3DRMMeshB
 
 static BOOL WINAPI d3drm_mesh_builder2_GetPerspective(IDirect3DRMMeshBuilder2 *iface)
 {
-    FIXME("iface %p stub!\n", iface);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder2(iface);
+
+    TRACE("iface %p.\n", iface);
 
-    return FALSE;
+    return mesh_builder->perspective;
 }
 
 static int WINAPI d3drm_mesh_builder2_GetFaceCount(IDirect3DRMMeshBuilder2 *iface)
@@ -1601,24 +1670,57 @@ static HRESULT WINAPI d3drm_mesh_builder3_Scale(IDirect3DRMMeshBuilder3 *iface,
 static HRESULT WINAPI d3drm_mesh_builder3_Translate(IDirect3DRMMeshBuilder3 *iface,
         D3DVALUE tx, D3DVALUE ty, D3DVALUE tz)
 {
-    FIXME("iface %p, tx %.8e, ty %.8e, tz %.8e stub!\n", iface, tx, ty, tz);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder3(iface);
+    DWORD i;
 
-    return E_NOTIMPL;
+    TRACE("iface %p, tx %.8e, ty %.8e, tz %.8e.\n", iface, tx, ty, tz);
+
+    for (i = 0; i < mesh_builder->nb_vertices; ++i)
+    {
+#ifndef SG_MUTANT_DRMMB_NO_TRANSLATE
+        mesh_builder->vertices[i].x += tx;
+#endif
+        mesh_builder->vertices[i].y += ty;
+        mesh_builder->vertices[i].z += tz;
+    }
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder3_SetColorSource(IDirect3DRMMeshBuilder3 *iface,
         D3DRMCOLORSOURCE source)
 {
-    FIXME("iface %p, source %#x stub!\n", iface, source);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder3(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, source %#x.\n", iface, source);
+
+    if ((int)source != D3DRMCOLOR_FROMFACE && (int)source != D3DRMCOLOR_FROMVERTEX)
+        return D3DRMERR_BADVALUE;
+    mesh_builder->color_source = source;
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder3_GetBox(IDirect3DRMMeshBuilder3 *iface, D3DRMBOX *box)
 {
-    FIXME("iface %p, box %p stub!\n", iface, box);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder3(iface);
+    DWORD i;
 
-    return E_NOTIMPL;
+    TRACE("iface %p, box %p.\n", iface, box);
+
+    if (!box)
+        return D3DRMERR_BADVALUE;
+    memset(box, 0, sizeof(*box));
+    for (i = 0; i < mesh_builder->nb_vertices; ++i)
+    {
+        const D3DVECTOR *v = &mesh_builder->vertices[i];
+
+        if (!i || v->x < box->min.x) box->min.x = v->x;
+        if (!i || v->y < box->min.y) box->min.y = v->y;
+        if (!i || v->z < box->min.z) box->min.z = v->z;
+        if (!i || v->x > box->max.x) box->max.x = v->x;
+        if (!i || v->y > box->max.y) box->max.y = v->y;
+        if (!i || v->z > box->max.z) box->max.z = v->z;
+    }
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder3_GenerateNormals(IDirect3DRMMeshBuilder3 *iface,
@@ -1631,9 +1733,11 @@ static HRESULT WINAPI d3drm_mesh_builder3_GenerateNormals(IDirect3DRMMeshBuilder
 
 static D3DRMCOLORSOURCE WINAPI d3drm_mesh_builder3_GetColorSource(IDirect3DRMMeshBuilder3 *iface)
 {
-    FIXME("iface %p stub!\n", iface);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder3(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p.\n", iface);
+
+    return mesh_builder->color_source;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder3_AddMesh(IDirect3DRMMeshBuilder3 *iface, IDirect3DRMMesh *mesh)
@@ -1678,10 +1782,19 @@ static HRESULT WINAPI d3drm_mesh_builder3_AddFaces(IDirect3DRMMeshBuilder3 *ifac
 static HRESULT WINAPI d3drm_mesh_builder3_ReserveSpace(IDirect3DRMMeshBuilder3 *iface,
         DWORD vertex_count, DWORD normal_count, DWORD face_count)
 {
-    FIXME("iface %p, vertex_count %lu, normal_count %lu, face_count %lu stub!\n",
-            iface, vertex_count, normal_count, face_count);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder3(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, vertex_count %lu, normal_count %lu, face_count %lu.\n", iface, vertex_count, normal_count, face_count);
+
+    if (vertex_count > mesh_builder->vertices_size &&
+            !d3drm_array_reserve((void **)&mesh_builder->vertices, &mesh_builder->vertices_size, vertex_count,
+            sizeof(*mesh_builder->vertices)))
+        return E_OUTOFMEMORY;
+    if (normal_count > mesh_builder->normals_size &&
+            !d3drm_array_reserve((void **)&mesh_builder->normals, &mesh_builder->normals_size, normal_count,
+            sizeof(*mesh_builder->normals)))
+        return E_OUTOFMEMORY;
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder3_SetColorRGB(IDirect3DRMMeshBuilder3 *iface,
@@ -1742,9 +1855,13 @@ static HRESULT WINAPI d3drm_mesh_builder3_SetMaterial(IDirect3DRMMeshBuilder3 *i
 static HRESULT WINAPI d3drm_mesh_builder3_SetTextureTopology(IDirect3DRMMeshBuilder3 *iface,
         BOOL wrap_u, BOOL wrap_v)
 {
-    FIXME("iface %p, wrap_u %#x, wrap_v %#x stub!\n", iface, wrap_u, wrap_v);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder3(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, wrap_u %#x, wrap_v %#x.\n", iface, wrap_u, wrap_v);
+
+    mesh_builder->wrap_u = !!wrap_u;
+    mesh_builder->wrap_v = !!wrap_v;
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder3_SetQuality(IDirect3DRMMeshBuilder3 *iface,
@@ -1762,25 +1879,44 @@ static HRESULT WINAPI d3drm_mesh_builder3_SetQuality(IDirect3DRMMeshBuilder3 *if
 static HRESULT WINAPI d3drm_mesh_builder3_SetPerspective(IDirect3DRMMeshBuilder3 *iface,
         BOOL enable)
 {
-    FIXME("iface %p, enable %#x stub!\n", iface, enable);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder3(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, enable %#x.\n", iface, enable);
+
+    mesh_builder->perspective = !!enable;
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder3_SetVertex(IDirect3DRMMeshBuilder3 *iface,
         DWORD index, D3DVALUE x, D3DVALUE y, D3DVALUE z)
 {
-    FIXME("iface %p, index %lu, x %.8e, y %.8e, z %.8e stub!\n", iface, index, x, y, z);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder3(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, index %lu, x %.8e, y %.8e, z %.8e.\n", iface, index, x, y, z);
+
+#ifndef SG_MUTANT_DRMMB_NO_BOUNDS
+    if (index >= mesh_builder->nb_vertices)
+        return D3DRMERR_BADVALUE;
+#endif
+    mesh_builder->vertices[index].x = x;
+    mesh_builder->vertices[index].y = y;
+    mesh_builder->vertices[index].z = z;
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder3_SetNormal(IDirect3DRMMeshBuilder3 *iface,
         DWORD index, D3DVALUE x, D3DVALUE y, D3DVALUE z)
 {
-    FIXME("iface %p, index %lu, x %.8e, y %.8e, z %.8e stub!\n", iface, index, x, y, z);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder3(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, index %lu, x %.8e, y %.8e, z %.8e.\n", iface, index, x, y, z);
+
+    if (index >= mesh_builder->nb_normals)
+        return D3DRMERR_BADVALUE;
+    mesh_builder->normals[index].x = x;
+    mesh_builder->normals[index].y = y;
+    mesh_builder->normals[index].z = z;
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder3_SetTextureCoordinates(IDirect3DRMMeshBuilder3 *iface,
@@ -1828,11 +1964,12 @@ static HRESULT WINAPI d3drm_mesh_builder3_GetGeometry(IDirect3DRMMeshBuilder3 *i
         DWORD *vertex_count, D3DVECTOR *vertices, DWORD *normal_count, D3DVECTOR *normals,
         DWORD *face_data_size, DWORD *face_data)
 {
-    FIXME("iface %p, vertex_count %p, vertices %p, normal_count %p, normals %p, "
-            "face_data_size %p, face_data %p stub!\n",
-            iface, vertex_count, vertices, normal_count, normals, face_data_size, face_data);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder3(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, vertex_count %p, vertices %p, normal_count %p, normals %p, face_data_size %p, face_data %p.\n", iface, vertex_count, vertices, normal_count, normals, face_data_size, face_data);
+
+    return IDirect3DRMMeshBuilder2_GetVertices(&mesh_builder->IDirect3DRMMeshBuilder2_iface, vertex_count, vertices,
+            normal_count, normals, face_data_size, face_data);
 }
 
 static HRESULT WINAPI d3drm_mesh_builder3_GetTextureCoordinates(IDirect3DRMMeshBuilder3 *iface,
@@ -1913,9 +2050,11 @@ static D3DRMRENDERQUALITY WINAPI d3drm_mesh_builder3_GetQuality(IDirect3DRMMeshB
 
 static BOOL WINAPI d3drm_mesh_builder3_GetPerspective(IDirect3DRMMeshBuilder3 *iface)
 {
-    FIXME("iface %p stub!\n", iface);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder3(iface);
+
+    TRACE("iface %p.\n", iface);
 
-    return FALSE;
+    return mesh_builder->perspective;
 }
 
 static int WINAPI d3drm_mesh_builder3_GetFaceCount(IDirect3DRMMeshBuilder3 *iface)
@@ -2087,17 +2226,27 @@ static HRESULT WINAPI d3drm_mesh_builder3_GetFace(IDirect3DRMMeshBuilder3 *iface
 static HRESULT WINAPI d3drm_mesh_builder3_GetVertex(IDirect3DRMMeshBuilder3 *iface,
         DWORD index, D3DVECTOR *vector)
 {
-    FIXME("iface %p, index %lu, vector %p stub!\n", iface, index, vector);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder3(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, index %lu, vector %p.\n", iface, index, vector);
+
+    if (index >= mesh_builder->nb_vertices || !vector)
+        return D3DRMERR_BADVALUE;
+    *vector = mesh_builder->vertices[index];
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder3_GetNormal(IDirect3DRMMeshBuilder3 *iface,
         DWORD index, D3DVECTOR *vector)
 {
-    FIXME("iface %p, index %lu, vector %p stub!\n", iface, index, vector);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder3(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, index %lu, vector %p.\n", iface, index, vector);
+
+    if (index >= mesh_builder->nb_normals || !vector)
+        return D3DRMERR_BADVALUE;
+    *vector = mesh_builder->normals[index];
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder3_DeleteVertices(IDirect3DRMMeshBuilder3 *iface,
@@ -2202,9 +2351,14 @@ static HRESULT WINAPI d3drm_mesh_builder3_AddTriangles(IDirect3DRMMeshBuilder3 *
 static HRESULT WINAPI d3drm_mesh_builder3_SetVertices(IDirect3DRMMeshBuilder3 *iface,
         DWORD start_idx, DWORD count, D3DVECTOR *vector)
 {
-    FIXME("iface %p, start_idx %lu, count %lu, vector %p stub!\n", iface, start_idx, count, vector);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder3(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, start_idx %lu, count %lu, vector %p.\n", iface, start_idx, count, vector);
+
+    if (!vector || start_idx > mesh_builder->nb_vertices || count > mesh_builder->nb_vertices - start_idx)
+        return D3DRMERR_BADVALUE;
+    memcpy(&mesh_builder->vertices[start_idx], vector, count * sizeof(*vector));
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder3_GetVertices(IDirect3DRMMeshBuilder3 *iface,
@@ -2227,10 +2381,14 @@ static HRESULT WINAPI d3drm_mesh_builder3_GetVertices(IDirect3DRMMeshBuilder3 *i
 static HRESULT WINAPI d3drm_mesh_builder3_SetNormals(IDirect3DRMMeshBuilder3 *iface,
         DWORD start_idx, DWORD count, D3DVECTOR *vector)
 {
-    FIXME("iface %p, start_idx %lu, count %lu, vector %p stub!\n",
-            iface, start_idx, count, vector);
+    struct d3drm_mesh_builder *mesh_builder = impl_from_IDirect3DRMMeshBuilder3(iface);
 
-    return E_NOTIMPL;
+    TRACE("iface %p, start_idx %lu, count %lu, vector %p.\n", iface, start_idx, count, vector);
+
+    if (!vector || start_idx > mesh_builder->nb_normals || count > mesh_builder->nb_normals - start_idx)
+        return D3DRMERR_BADVALUE;
+    memcpy(&mesh_builder->normals[start_idx], vector, count * sizeof(*vector));
+    return D3DRM_OK;
 }
 
 static HRESULT WINAPI d3drm_mesh_builder3_GetNormals(IDirect3DRMMeshBuilder3 *iface,
